string/string.cc

```diff
@@ -1,12 +1,154 @@
 #include "string.h"
 
-String::String()
+#include <algorithm>
+#include <iterator>
+#include <memory>
+#include <string>
+#include <cstring>
+
+const char* EMPTY = "";
+
+String::String() = default;
+
+String::~String() = default;
+
+String::String(String&& o) = default;
+
+String::String(const String& o): String(o.begin(), o.end()) {}
+
+String::String(const char* s): String(s, s + std::strlen(s)) {}
+
+template String::String(std::string::iterator first, std::string::iterator last);
+
+template<typename It> String::String(It first, It last)
+{
+	mSize = std::distance(first, last);
+	mValue.reset(new char[mSize + 1]);
+
+	std::copy_n(first, mSize, begin());
+
+	mValue[mSize + 1] = '\0';
+}
+
+// I don't see any other effective way to concatenate for the given API.
+String::String(const String& a, const String& b)
```

Comment:
Is there a smart concatenation without copying each time?
In Java all  '+' are converted to a StringBuffer by compiler. What's the point of '+' if you need to use a buffer anyway?